### PR TITLE
[junit5] initial

### PR DESCRIPTION
**Pull request by rotty3000**
Signed-off-by: Raymond Augé <raymond.auge@liferay.com>

### CODE REVIEW

**Comment by rotty3000** [comment]
hey @bjhargrave this PR contains a profile called `bnd-next` which uses the latest bnd snapshot.

`mvn clean install -Pbnd-next`

I wonder if once this clears we could add a travis job to build this.

**Comment by rotty3000** [comment]
btw @kriegfrj , this also contains the bnd tester.junit-platform junit5 and the peer extensions for bundle context and service use along with all the same tests we have against junit4.

**Comment by rotty3000** [comment]
please note the junit5 tests only run when the `-Pbnd-next` profile is enabled.

**Comment by rotty3000** [comment]
I'd like to get this in since I think it's more or less stable. The main issues have been mostly in the junit5 test support which seems to be stabilizing in `bndtools` and `biz.aQute.tester.junit-pltform`.

Actually, getting this merged will give us a platform for stabilizing that story.

**Comment by kriegfrj** [comment]
I'll try and prioritize this tomorrow.

File: org.osgi.test.junit5/src/test/java/org/osgi/test/junit5/ExecutorExtension.java [hunk 1]
```diff
@@ -0,0 +1,107 @@
...
+import org.junit.jupiter.api.extension.ExtensionConfigurationException;
+import org.junit.jupiter.api.extension.ExtensionContext;
+import org.junit.jupiter.api.extension.ExtensionContext.Namespace;
+import org.junit.jupiter.api.extension.ParameterContext;
+import org.junit.jupiter.api.extension.ParameterResolutionException;
+import org.junit.jupiter.api.extension.ParameterResolver;
+import org.junit.platform.commons.util.ExceptionUtils;
+import org.junit.platform.commons.util.ReflectionUtils;
+
+public class ExecutorExtension implements AfterEachCallback, BeforeEachCallback, ParameterResolver {
```
Comment (kriegfrj) [inline org.osgi.test.junit5/src/test/java/org/osgi/test/junit5/ExecutorExtension.java]: Although it perhaps goes beyond the scope of osgi-test, there is a lot about proper osgi testing that involves concurrency. I think we should consider making these executor-related extensions/rules part of the public API. There are also some concurrency-related utilities I wrote as part of the bnd tester testing that have generic application (eg, waiting for another thread to enter the wait state) that could form part of a concurrency testing library. That can be a topic for another issue/PR though; happy for it to be test-only scope for now.

File: org.osgi.test.junit5/src/test/java/org/osgi/test/junit5/service/ServiceUseExtensionTest.java [hunk 2]
```diff
@@ -0,0 +1,359 @@
...
+import org.mockito.stubbing.Answer;
+import org.osgi.framework.InvalidSyntaxException;
+import org.osgi.test.common.tracking.TrackServices;
+import org.osgi.test.junit5.ExecutorExtension;
+import org.osgi.test.junit5.ExecutorParameter;
+import org.osgi.test.junit5.types.Foo;
+import org.osgi.test.junit5.types.MockStore;
+
+@ExtendWith(ExecutorExtension.class)
+public class ServiceUseExtensionTest {
```
Comment (kriegfrj) [inline org.osgi.test.junit5/src/test/java/org/osgi/test/junit5/service/ServiceUseExtensionTest.java]: If `ServiceUseRule` and `ServiceUseExtension` had a common base class, then their respective test classes could too. That way you can avoid duplicating the test code between each and also give a greater guarantee that their behaviour is consistent.

File: org.osgi.test.common/src/main/java/org/osgi/test/common/service/ServiceAware.java [hunk 3]
```diff
@@ -0,0 +1,48 @@
...
+package org.osgi.test.common.service;
+
+import java.util.List;
+import java.util.SortedMap;
+
+import org.osgi.annotation.versioning.ProviderType;
+import org.osgi.framework.ServiceReference;
+
+@ProviderType
+public interface ServiceAware<T> {
```
[the remark this answers is not in the document]
Comment (rotty3000) [reply org.osgi.test.common/src/main/java/org/osgi/test/common/service/ServiceAware.java]: I took your advice. see updated

File: org.osgi.test.common/src/main/java/org/osgi/test/common/tracking/TrackingConfig.java [hunk 4]
```diff
@@ -0,0 +1,31 @@
...
+ * limitations under the License.
+ */
+
+package org.osgi.test.common.tracking;
+
+import org.osgi.annotation.versioning.ProviderType;
+import org.osgi.framework.Filter;
+
+@ProviderType
+public interface TrackingConfig {
```
[the remark this answers is not in the document]
Comment (rotty3000) [reply org.osgi.test.common/src/main/java/org/osgi/test/common/tracking/TrackingConfig.java]: same

File: org.osgi.test.junit4/src/main/java/org/osgi/test/junit4/service/ServiceUseRule.java [hunk 5]
```diff
@@ -109,15 +114,16 @@ public Builder(Class<T> serviceType, BundleContextRule bundleContextRule) {
 
 	}
 
-	private final ServiceUse<T> use;
-	private final Class<T>			serviceType;
 	private final BundleContextRule	bundleContextRule;
+	private final T[]				empty;
```
[the remark this answers is not in the document]
Comment (rotty3000) [reply org.osgi.test.junit4/src/main/java/org/osgi/test/junit4/service/ServiceUseRule.java]: done!

File: org.osgi.test.junit5/src/test/java/org/osgi/test/junit5/ExecutorExtension.java (hunk 1 above)
Comment (rotty3000) [reply org.osgi.test.junit5/src/test/java/org/osgi/test/junit5/ExecutorExtension.java]: issue created.

File: org.osgi.test.junit5/src/test/java/org/osgi/test/junit5/service/ServiceUseExtensionTest.java (hunk 2 above)
Comment (rotty3000) [reply org.osgi.test.junit5/src/test/java/org/osgi/test/junit5/service/ServiceUseExtensionTest.java]: that would have to be a new module right? You can't share code from a module that has it in `src/test/main` right? I think that's wonky!?! I'm open to ideas.

Comment (rotty3000) [reply org.osgi.test.junit5/src/test/java/org/osgi/test/junit5/service/ServiceUseExtensionTest.java]: The other issue is exposing the junit apis to each other. I'm happy keeping the code somewhat duplicative in that there is no junit 3/4 api exposure to junit5 testing and vice-versa.

Comment (rotty3000) [reply org.osgi.test.junit5/src/test/java/org/osgi/test/junit5/service/ServiceUseExtensionTest.java]: I filed an issue for this.

**Pull request merged**